### PR TITLE
[Resolve #1468] Cast parameters to strings

### LINKED ISSUE #1468: Reimplement #1443 - parameter Schema Validation
### Subject of the issue
In already-resolved and merged issue #1443  we implemented parameter validation. In Slack, we then discussed a preferred implementation of using instead a parameter Schema Validator.

This issue tracks possible reimplementation of this feature.

### Your environment
* version of sceptre (sceptre --version) latest
* version of python (python --version) 3.10
* which OS/distro Ubuntu

### More info
```
% git show f3b7e51fee1ff78a457f20b5dddac91254a5fdf2 
```
- Merged PR https://github.com/Sceptre/sceptre/pull/1442
- Slack discussion is in the archives.
- Schema validator suggested was pydantic.

### PR DESCRIPTION
**Pull request by alexharv074**
In already-resolved and merged issue #1443 we implemented parameter validation.

Subsequent discussion led to a consensus that it would be better to not only validate parameters, but also cast where applicable to string. This will lead to a cleaner user experience for most Sceptre users where parameters passed as ints, floats and bools in their YAML values files are cast to strings first and then sent to CloudFormation, since CloudFormation's API is unable to handle other types.

This PR changes the `_ensure_parameters` method to `_cast_parameters` in the Stack class to handle casting to strings where applicable instead of only validating.

## PR Checklist

- [x] Wrote a good commit message & description [see guide below].
- [x] Commit message starts with `[Resolve #issue-number]`.
- [x] Added/Updated unit tests.
- [ ] Added/Updated integration tests (if applicable).
- [x] All unit tests (`poetry run tox`) are passing.
- [x] Used the same coding conventions as the rest of the project.
- [x] The new code passes pre-commit validations (`poetry run pre-commit run --all-files`).
- [x] The PR relates to _only_ one subject with a clear title.
      and description in grammatically correct, complete sentences.

## Approver/Reviewer Checklist

- [ ] Before merge squash related commits.

## Other Information

[Guide to writing a good commit](http://chris.beams.io/posts/git-commit/)

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -28,7 +28,7 @@ repos:
         exclude: |
           (?x)(
             ^integration-tests/sceptre-project/config/|
-            ^integration-tests/sceptre-project/templates/jinja/|
+            ^integration-tests/sceptre-project/templates/|
```
Comment (alexharv074) [inline .pre-commit-config.yaml:31]: These changes are necessary from this PR https://github.com/Sceptre/sceptre/pull/1470 I merged that branch in here so that the build could be green.

File: sceptre/stack.py [hunk 2]
```diff
@@ -281,6 +281,17 @@ def _ensure_parameters(
     ) -> Dict[str, Union[str, List[Union[str, Resolver]], Resolver]]:
         """Ensure CloudFormation parameters are of valid types"""
 
+        def cast_value(value: Any) -> Union[str, List[Union[str, Resolver]], Resolver]:
```
Comment (zaro0508) [inline sceptre/stack.py:284]: This function's purpose was only to validate parameters, as indicated in this methods docstring.  Could we update the docstring to reflect that this method is also casting parameters?  Or what might be even more clear is to split validation and casting actions into two separate methods?

**Review by zaro0508 — CHANGES_REQUESTED** [review]
generally LGTM but would like the method to be more concise and clear.

File: sceptre/stack.py (hunk 2 above)
Comment (alexharv074) [reply sceptre/stack.py:284]: @zaro0508 I see your point. On looking at this, splitting it up into 2 or 3 methods feels messier to me than sticking with inner functions but fixing the naming. To avoid the inconsistency between `_ensure_boolean` and `_ensure_parameters`, which now would do different things, I've renamed as `_cast_parameters` and fixed the docstring. What do think about this version?

**Comment by alexharv074** [comment]
> the PR for that issue was not merged?
> 

It was merged in f3b7e51fee1ff78a457f20b5dddac91254a5fdf2 but we have never released those changes thus no users will be impacted by the change (unless they are consuming the master branch).

**Review by zaro0508 — APPROVED** [review]
thanks for the clarification

**Comment by fen9li** [comment]
Hey Alex,

Good day. I received an error may have something to do with this merge.

My `config/bla-bla/config.yaml` file look like below:

```
---
sceptre_role: <aws-role-arn>

environment: bla-bla
application: bla-bla
accountid: <aws-account-id>
instance_type: t3.large
db_instance_type: db.t3.small
license_model: license-included
db_storage: '40'
multi_az: 'false'
restore_db_snapshot_identifier: 'null'

stack_tags_env:
  xxxxxx: xxxxxx
```

The error:

```
[2024-09-09T12:43:16.398Z] + sceptre --var-file config/config.yaml --var-file config/bla-bla/config.yaml launch -y bla-bla

[2024-09-09T12:43:16.964Z] "bla-bla/rds-instances: Values for parameters must be strings, lists or resolvers, got {'Environment': 'bla-bla', 'Application': 'bla-bla', 'RDSMasterUser': <username>, 'RDSMasterUserPassword': <password>, 'InstanceType': 'db.t3.small', 'LicenseModel': 'license-included', 'AllocatedStorageGB': '40', 'MultiAZ': 'false', 'RestoreDBSnapshotIdentifier': None}"

script returned exit code 1
```

I reckon that the string 'null' in my `config/bla-bla/config.yaml` somehow converted to None and failed the new added validation. Any chance to add one more test `            {"NullAsString": "null"},` here: 

```
 "parameters",
        [
            {"IntAsString": "1"},
            {"Int": 1},
            {"BoolAsString": "true"},
            {"Bool": True},
            {"List": ["aStringAndA", FakeResolver()]},
            {"ListOfInts": [1, 2, 3]},
            {"Resolver": FakeResolver()},
        ],
```

So we can know if there is any bug here? Thank you very much.
Feng

**Comment by fen9li** [comment]
Hey Alex,

I fixed the issue already. Need adding "" to the value when passed down like below.

```
 RestoreDBSnapshotIdentifier: "{{ var.restore_db_snapshot_identifier }}"
```

Thanks.
Feng
